File: app/src/main/java/org/wikipedia/analytics/ToCInteractionFunnel.java
package org.wikipedia.analytics;

import androidx.annotation.NonNull;

import org.json.JSONObject;
import org.wikipedia.WikipediaApp;
import org.wikipedia.dataclient.WikiSite;

import java.util.concurrent.TimeUnit;

public class ToCInteractionFunnel extends Funnel {
    private static final String SCHEMA_NAME = "MobileWikiAppToCInteraction";
    private static final int REV_ID = 19044853;

    private final int pageId;
    private final int numSections;

    private int numOpens;
    private int numSectionClicks;

    private long lastScrollStartMillis;
    private int totalOpenedSec;

    public ToCInteractionFunnel(WikipediaApp app, WikiSite wiki, int pageId, int numSections) {
        super(app, SCHEMA_NAME, REV_ID, Funnel.SAMPLE_LOG_ALL, wiki);
        this.pageId = pageId;
        this.numSections = numSections;
    }

    @Override
    protected JSONObject preprocessData(@NonNull JSONObject eventData) {
        preprocessData(eventData, "page_id", pageId);
        preprocessData(eventData, "num_sections", numSections);
        return super.preprocessData(eventData);
    }

    @Override protected void preprocessSessionToken(@NonNull JSONObject eventData) { }

    public void scrollStart() {
        numOpens++;
        lastScrollStartMillis = System.currentTimeMillis();
    }

    public void scrollStop() {
        if (lastScrollStartMillis == 0) {
            return;
        }
        totalOpenedSec += (System.currentTimeMillis() - lastScrollStartMillis) / TimeUnit.SECONDS.toMillis(1);
        lastScrollStartMillis = 0;
    }


    public void logClick() {
        numSectionClicks++;
    }

    public void log() {
        scrollStop();
        if (numSections == 0 || numOpens == 0) {
            return;
        }
        log(
                "num_peeks", 0,
                "num_opens", numOpens,
                "num_section_clicks", numSectionClicks,
                "total_peek_sec", 0,
                "total_open_sec", totalOpenedSec
        );
    }
}

File: app/src/main/java/org/wikipedia/analytics/ToCInteractionFunnel.kt
package org.wikipedia.analytics

import org.json.JSONObject
import org.wikipedia.WikipediaApp
import org.wikipedia.dataclient.WikiSite
import java.util.concurrent.TimeUnit

class ToCInteractionFunnel(app: WikipediaApp, wiki: WikiSite?, private val pageId: Int, private val numSections: Int) :
        Funnel(app, SCHEMA_NAME, REV_ID, SAMPLE_LOG_ALL, wiki) {

    private var numOpens = 0
    private var numSectionClicks = 0
    private var lastScrollStartMillis: Long = 0
    private var totalOpenedSec = 0

    override fun preprocessData(eventData: JSONObject): JSONObject? {
        preprocessData(eventData, "page_id", pageId)
        preprocessData(eventData, "num_sections", numSections)
        return super.preprocessData(eventData)
    }

    override fun preprocessSessionToken(eventData: JSONObject) {}

    fun scrollStart() {
        numOpens++
        lastScrollStartMillis = System.currentTimeMillis()
    }

    fun scrollStop() {
        if (lastScrollStartMillis == 0L) {
            return
        }
        totalOpenedSec += ((System.currentTimeMillis() - lastScrollStartMillis) / TimeUnit.SECONDS.toMillis(1)).toInt()
        lastScrollStartMillis = 0
    }

    fun logClick() {
        numSectionClicks++
    }

    fun log() {
        scrollStop()
        if (numSections == 0 || numOpens == 0) {
            return
        }
        log(
                "num_peeks", 0,
                "num_opens", numOpens,
                "num_section_clicks", numSectionClicks,
                "total_peek_sec", 0,
                "total_open_sec", totalOpenedSec
        )
    }

    companion object {
        private const val SCHEMA_NAME = "MobileWikiAppToCInteraction"
        private const val REV_ID = 19044853
    }
}
